[Clang][DebugInfo] Clang generates an extra spurious unnamed 'dbg.declare'

CarlosAlbertoEnciso opened the pull request:
Do not emit call to llvm.dbg.declare when the variable declaration
is a DecompositionDecl as its instance class is always unnamed.

The emitted debug declare looks like:
```
call void @llvm.dbg.declare(metadata ..., metadata !xx, metadata ...)
!xx = !DILocalVariable(scope: !..., file: !..., line: ..., type: !...)
```

CarlosAlbertoEnciso committed:
[Clang][DebugInfo] Clang generates an extra spurious unnamed 'dbg.declare'.

Do not emit call to llvm.dbg.declare when the variable declaration
is a DecompositionDecl as its instance class is always unnamed.

The emitted debug declare looks like:

  call void @llvm.dbg.declare(metadata ..., metadata !xx, metadata ...)
  !xx = !DILocalVariable(scope: !..., file: !..., line: ..., type: !...)

OCHyams commented on clang/test/CodeGenCXX/debug-info-structured-binding-field.cpp:
```diff
@@ -0,0 +1,20 @@
+// CHECK-LABEL: define dso_local noundef i32 @main
+// CHECK:      %retval = alloca i32, align 4
+// CHECK-NEXT: [[T0:%.*]] = alloca %struct.Tuple, align 4
+// CHECK:      call void @llvm.dbg.declare(metadata ptr [[T0]], metadata {{.*}}, metadata !DIExpression())
+// CHECK:      call void @llvm.dbg.declare(metadata ptr [[T0]], metadata {{.*}}, metadata !DIExpression(DW_OP_plus_uconst, {{[0-9]+}}))
+// CHECK-NOT:  call void @llvm.dbg.declare(metadata ptr [[T0]], metadata {{.*}}, metadata !DIExpression())
```
It doesn't look like this test covers anything that `clang/test/CodeGenCXX/debug-info-structured-binding.cpp` doesn't already?

IMO no need to add this one. Please can you modify the other test to check that the DILocalVariables in the dbg.declares have the expected names (i.e., confirm there's no unnamed variable).

OCHyams commented on clang/test/CodeGenCXX/debug-info-structured-binding.cpp:
```diff
@@ -2,10 +2,8 @@
 
```
Please can you add `--implicit-check-not="call void @llvm.dbg.declare"` to the `FileCheck` command?

CarlosAlbertoEnciso replied:
Added the `--implicit-check-not` option.

OCHyams reviewed (commented):
This SGTM.

I don't think the unnamed variable serves any useful purpose but may not being imaginative enough. I can't see any discussion on the topic when the code was added in [D119178](https://reviews.llvm.org/D119178) . CC the author, @shafik (or @adrian-prantl) - is there a use-case for the anonymous variable that covers the whole of the structured binding storage?

adrian-prantl reviewed (approved):
Aside from other open comments the spirit of the change LGTM.

CarlosAlbertoEnciso committed:
[Clang][DebugInfo] Clang generates an extra spurious unnamed 'dbg.declare'.

Do not emit call to llvm.dbg.declare when the variable declaration
is a DecompositionDecl as its instance class is always unnamed.

The emitted debug declare looks like:

  call void @llvm.dbg.declare(metadata ..., metadata !xx, metadata ...)
  !xx = !DILocalVariable(scope: !..., file: !..., line: ..., type: !...)

- Remove the added test case.
- Added the '--implicit-check-not' option
- Added checks for variables names.

CarlosAlbertoEnciso committed:
[Clang][DebugInfo] Clang generates an extra spurious unnamed 'dbg.declare'.

Do not emit call to llvm.dbg.declare when the variable declaration
is a DecompositionDecl as its instance class is always unnamed.

The emitted debug declare looks like:
  call void @llvm.dbg.declare(metadata ..., metadata !xx, metadata ...)
  !xx = !DILocalVariable(scope: !..., file: !..., line: ..., type: !...)

- Added check for missing captured 'a' variable.
- Check for the specific values in DW_OP_plus_uconst.

OCHyams reviewed (approved):
New changes LGTM, thanks!

CarlosAlbertoEnciso commented:
@adrian-prantl, @OCHyams Thanks for your reviews.
